File: repo/backend/internal/model/feedback.go
package model

import "time"

// FeedbackTag is a controlled-vocabulary label applied to feedback.
type FeedbackTag struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	IsActive bool   `json:"is_active"`
}

// Feedback is a star rating and optional tagged comment from a reader.
type Feedback struct {
	ID          string     `json:"id"`
	BranchID    string     `json:"branch_id"`
	ReaderID    string     `json:"reader_id"`
	TargetType  string     `json:"target_type"` // holding, program
	TargetID    string     `json:"target_id"`
	Rating      *int       `json:"rating,omitempty"` // 1–5
	Comment     *string    `json:"comment,omitempty"`
	Status      string     `json:"status"` // pending, approved, rejected, flagged
	ModeratedBy *string    `json:"moderated_by,omitempty"`
	ModeratedAt *time.Time `json:"moderated_at,omitempty"`
	SubmittedAt time.Time  `json:"submitted_at"`
	Tags        []string   `json:"tags,omitempty"` // populated by join, not stored directly
}

// Appeal is a reader-submitted challenge to a decision.
type Appeal struct {
	ID          string    `json:"id"`
	BranchID    string    `json:"branch_id"`
	ReaderID    string    `json:"reader_id"`
	AppealType  string    `json:"appeal_type"` // enrollment_denial, account_suspension, feedback_rejection, blacklist_removal, other
	TargetType  *string   `json:"target_type,omitempty"`
	TargetID    *string   `json:"target_id,omitempty"`
	Reason      string    `json:"reason"`
	Status      string    `json:"status"` // submitted, under_review, resolved, dismissed
	SubmittedAt time.Time `json:"submitted_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// AppealArbitration is the formal decision record for an appeal.
type AppealArbitration struct {
	ID             string    `json:"id"`
	AppealID       string    `json:"appeal_id"`
	ArbitratorID   string    `json:"arbitrator_id"`
	Decision       string    `json:"decision"` // upheld, dismissed, partial
	DecisionNotes  string    `json:"decision_notes"`
	BeforeState    any       `json:"before_state,omitempty"` // JSONB snapshot
	AfterState     any       `json:"after_state,omitempty"`  // JSONB snapshot
	DecidedAt      time.Time `json:"decided_at"`
}

File: repo/frontend/src/api/feedback.ts
// API client for reader feedback.

import { apiClient } from './client';
import type { PageResult } from './readers';

export interface FeedbackTag {
  id: string;
  name: string;
  is_active: boolean;
}

export interface Feedback {
  id: string;
  branch_id: string;
  reader_id: string;
  target_type: 'holding' | 'program';
  target_id: string;
  rating?: number;
  comment?: string;
  tags: string[];
  status: 'pending' | 'approved' | 'rejected' | 'flagged';
  moderated_by?: string;
  moderated_at?: string;
  submitted_at: string;
}

function listFeedback(params?: {
  status?: string;
  target_type?: string;
  target_id?: string;
  page?: number;
  per_page?: number;
}): Promise<PageResult<Feedback>> {
  const qs = new URLSearchParams();
  if (params?.status) qs.set('status', params.status);
  if (params?.target_type) qs.set('target_type', params.target_type);
  if (params?.target_id) qs.set('target_id', params.target_id);
  if (params?.page) qs.set('page', String(params.page));
  if (params?.per_page) qs.set('per_page', String(params.per_page));
  const q = qs.toString();
  return apiClient.get<PageResult<Feedback>>(`/feedback${q ? '?' + q : ''}`);
}

function getFeedback(id: string): Promise<Feedback> {
  return apiClient.get<Feedback>(`/feedback/${id}`);
}

function listTags(): Promise<FeedbackTag[]> {
  return apiClient.get<FeedbackTag[]>('/feedback/tags');
}

function submitFeedback(data: {
  reader_id: string;
  target_type: 'holding' | 'program';
  target_id: string;
  rating?: number;
  comment?: string;
  tags?: string[];
}): Promise<Feedback> {
  return apiClient.post<Feedback>('/feedback', { ...data, tags: data.tags ?? [] });
}

function moderateFeedback(id: string, status: 'approved' | 'rejected' | 'flagged'): Promise<Feedback> {
  return apiClient.post<Feedback>(`/feedback/${id}/moderate`, { status });
}

export const feedbackApi = {
  list: listFeedback,
  get: getFeedback,
  listTags,
  submit: submitFeedback,
  moderate: moderateFeedback,
};
